sled-agent/src/rack_setup/service.rs

```diff
@@ -371,36 +330,52 @@ impl ServiceInner {
         Ok(())
     }
 
-    // Configure the internal DNS servers with the initial DNS data
-    async fn initialize_dns(
+    // Ensure that all services of a particular type are running.
+    //
+    // This is useful in a rack-setup context, where initial boot ordering
+    // can matter for first-time-setup.
+    //
+    // Note that after first-time setup, the initialization order of
+    // services should not matter.
+    async fn ensure_all_services_of_type(
         &self,
         service_plan: &ServicePlan,
+        zone_types: &HashSet<std::mem::Discriminant<ZoneType>>,
```

Comment:
I'm a bit confused by the use of `Discriminant` here. Does `ZoneType` carry data? If not, I think it'd be less surprising to store the enum variants directly in the set.

Reply:
I wish I could! But [enum variants are not types](https://github.com/rust-lang/rfcs/pull/1450).

I really just wanted a way to say "accept some subset of the enum" -- `std::mem::discriminant` uniquely identifies enum variants, so this seemed to fit.

Reply:
I'm confused, too.  Can't you have a `HashSet<ZoneType>`?  That would allow any variant to appear in the hashset, but so does `HashSet<Discriminant<ZoneType>>`, doesn't it?

Reply:
Yeah, that makes sense, and is why I was asking about the variants containing data. I guess there are a few things. First, given that they _don't_ currently carry data, it seems preferable to me to use the enum directly as the value for the map. Second, even if we _do_ add data, do we really not want to distinguish between them in this code?

Reply:
I went ahead and swapped it back to a non-discriminant version in 96f3afd40 - maybe we can just revisit that question if we ever need to add data to this enum in the future?